File: core/src/main/java/y2k/joyreactor/common/Messenger.java
package y2k.joyreactor.common;

import rx.functions.Action1;

import java.util.*;

/**
 * Created by y2k on 01/10/15.
 */
public class Messenger {

    private static final Messenger sInstance = new Messenger();

    private Map<Class, Observable> observers = new HashMap<>();
    private Map<ActionObserver, Object> registrations = new HashMap<>();

    public void send(Object message) {
        Observable observable = observers.get(message.getClass());
        if (observable != null)
            observable.notifyObservers(message);
    }

    public <T> void register(Object receiver, Action1<T> callback, Class<T> type) {
        Observable observable = observers.get(type);
        if (observable == null) observers.put(type, observable = new ObservableImpl());

        ActionObserver<T> o = new ActionObserver<>(callback);
        observable.addObserver(o);

        registrations.put(o, receiver);
    }

    public void unregister(Object receiver) {
        Set<ActionObserver> keys = registrations.keySet();
        for (ActionObserver key : keys.toArray(new ActionObserver[keys.size()]))
            if (registrations.get(key) == receiver) {
                registrations.remove(key);
                for (Observable o : observers.values())
                    o.deleteObserver(key);
            }
    }

    public static Messenger getInstance() {
        return sInstance;
    }

    private static class ActionObserver<T> implements Observer {

        private Action1<T> callback;

        public ActionObserver(Action1<T> callback) {
            this.callback = callback;
        }

        @Override
        public void update(Observable o, Object arg) {
            callback.call((T) arg);
        }
    }

    private static class ObservableImpl extends Observable {

        @Override
        public void notifyObservers(Object arg) {
            setChanged();
            super.notifyObservers(arg);
        }
    }
}

File: core/src/main/kotlin/y2k/joyreactor/common/Messenger.kt
package y2k.joyreactor.common

import java.util.*

/**
 * Created by y2k on 01/10/15.
 */
object Messenger {

    private val observers = HashMap<Class<*>, Observable>()
    private val registrations = HashMap<ActionObserver<*>, Any>()

    fun send(message: Any) {
        val observable = observers[message.javaClass]
        observable?.notifyObservers(message)
    }

    fun <T> register(receiver: Any, callback: (T) -> Unit, type: Class<T>) {
        var observable = observers[type]
        if (observable == null) {
            observable = ObservableImpl()
            observers.put(type, observable)
        }

        val o = ActionObserver(callback)
        observable.addObserver(o)
        registrations.put(o, receiver)
    }

    fun unregister(receiver: Any) {
        for (key in registrations.keys.toList())
            if (registrations[key] === receiver) {
                registrations.remove(key)
                for (o in observers.values)
                    o.deleteObserver(key)
            }
    }

    private class ActionObserver<T>(private val callback: (T) -> Unit) : Observer {

        @Suppress("UNCHECKED_CAST")
        override fun update(o: Observable, arg: Any) {
            callback(arg as T)
        }
    }

    private class ObservableImpl : Observable() {

        override fun notifyObservers(arg: Any) {
            setChanged()
            super.notifyObservers(arg)
        }
    }
}